src/services/hardware/hardwareService.ts

```diff
@@ -0,0 +1,321 @@
+      throw new Error(`Lost connection to database (hub)! ${err}`);
+    }
+  };
+
+  /**
+   * Using the unique token the item is returned to the hardware library
+   * @param hardwareItemID
+   * @param token
+   */
+  itemToBeReturnedToLibrary = async (userID: number, hardwareItemID: number, token: string, itemQuantity: number): Promise<void> => {
+    // The item is being returned
+    try {
+      // Decrement the reservation count for the hardware item
+      await this.hardwareRepository
+        .decrement({ id: hardwareItemID }, "takenStock", itemQuantity);
+
+      // Delete the user reservation from the database
+      await this.reservedHardwareService.removeReservation(token);
+
+      const message: string = `UID ${userID} returned ${itemQuantity} of ${hardwareItemID}`;
+      new QueryLogger().hardwareLog(LoggerLevels.LOG, message);
+    } catch (err) {
+      throw new Error(`Lost connection to database (hub)! ${err}`);
+    }
+  };
+
+  /**
+   * Fetches all hardware items with their reservations
+   */
+  getAllHardwareItemsWithReservations = async (): Promise<HardwareItem[]> => {
+    const items: HardwareItem[] = await this.hardwareRepository
+      .createQueryBuilder("hardwareItem")
+      .leftJoinAndSelect("hardwareItem.reservations", "reservations")
+      .leftJoinAndSelect("reservations.user", "user")
+      .orderBy("hardwareItem.name")
+      .getMany();
+
+    return items;
+  };
+
+  /**
+   * Returns all the hardware items from the database in a formatted array
+   */
+  getAllHardwareItems = async (userId?: number): Promise<Object[]> => {
+    const hardwareItems: HardwareItem[] = await this.hardwareRepository
+      .find({
+        order: {
+          name: "ASC"
+        }
+      });
+
+    const allUserReservations: ReservedHardwareItem[] = await this.reservedHardwareService.getAllReservations();
+
+    const formattedData = [];
+    for (const item of hardwareItems) {
+      let remainingItemCount: number = item.totalStock - (item.reservedStock + item.takenStock);
+
+      // Check if any reservation for the current item have expired
+      const reservationsForItem: ReservedHardwareItem[] = allUserReservations.filter((reservation) => reservation.hardwareItem.name === item.name);
+      for (const itemReservation of reservationsForItem) {
+        if (itemReservation && itemReservation.isReserved && !this.isReservationValid(itemReservation.reservationExpiry)) {
+          remainingItemCount += itemReservation.reservationQuantity;
+          await this.reservedHardwareService.deleteReservation(itemReservation.reservationToken);
+        }
+      }
+      const userReservation: ReservedHardwareItem = allUserReservations.find((reservation) => reservation.hardwareItem.name === item.name && reservation.user.id === userId);
+
+      const isUsersReservation: boolean = (userReservation !== undefined && userId && userReservation.user.id === userId)
+        ? true : false;
+
+      formattedData.push({
```

Comment:
I think it would be a better idea to do this re-mapping somewhere else (maybe the controller or the view?). A service should always return an object that has a standard structure (i.e.: the model of the object in the DB) because the service is supposed to be reusable by various components of the app and using this custom structure limits the use of this method to this one specific case it's being used for now.

Reply:
That whole function needs some re-working. I'm not happy with it at the moment. I think I will do the remapping in the controller, plus it will make the function in the service more generic.

Reply:
I have re-written this function, and I think it is much cleaner now. I have moved quite a lot of the logic from this function into the `reservedHardwareService`.